File: Amazing Bird/Classes/Sprites/ABCloud.m
//
//  ABCloud.m
//  Amazing Bird
//
//  Created by ChristianEnevoldsen on 25/02/14.
//  Copyright (c) 2014 ChristianEnevoldsen. All rights reserved.
//

#import "ABCloud.h"
#import "ABGameScene.h"
#import "ABNormalBird.h"

@implementation ABCloud

- (instancetype)initAtPosition:(CGPoint)position {
	self = [super initWithTexture:[SKTexture textureWithImageNamed:@"Cloud"]
	                     position:position];
	if (self) {
		self.name = @"cloud";
	}
	return self;
}

- (void)configurePhysicsBody {
	CGFloat offsetX = self.size.width * self.anchorPoint.x;
	CGFloat offsetY = self.size.height * self.anchorPoint.y;

	CGMutablePathRef path = CGPathCreateMutable();

	if (UI_USER_INTERFACE_IDIOM() == UIUserInterfaceIdiomPad) {
		CGPathMoveToPoint(path, NULL, 17 - offsetX, 85 - offsetY);
		CGPathAddLineToPoint(path, NULL, 128 - offsetX, 91 - offsetY);
		CGPathAddLineToPoint(path, NULL, 147 - offsetX, 29 - offsetY);
		CGPathAddLineToPoint(path, NULL, 101 - offsetX, 2 - offsetY);
		CGPathAddLineToPoint(path, NULL, 6 - offsetX, 21 - offsetY);
	}
	else {
		CGPathMoveToPoint(path, NULL, 4 - offsetX, 45 - offsetY);
		CGPathAddLineToPoint(path, NULL, 70 - offsetX, 53 - offsetY);
		CGPathAddLineToPoint(path, NULL, 83 - offsetX, 14 - offsetY);
		CGPathAddLineToPoint(path, NULL, 46 - offsetX, 1 - offsetY);
		CGPathAddLineToPoint(path, NULL, 2 - offsetX, 14 - offsetY);
	}

	CGPathCloseSubpath(path);
	self.physicsBody = [SKPhysicsBody bodyWithPolygonFromPath:path];
	CGPathRelease(path);

	self.physicsBody.affectedByGravity = NO;
	self.physicsBody.linearDamping = 0;
	self.physicsBody.allowsRotation = NO;
	self.physicsBody.mass = 1e1;

	self.physicsBody.categoryBitMask = ABColliderTypeCloud;
	self.physicsBody.collisionBitMask = ABColliderTypeBird;
	self.physicsBody.contactTestBitMask = ABColliderTypeBird | ABColliderTypeCloud;
}

- (void)reset {
	[super reset];
    self.alpha = 1.0;
	collidedWithCloud = NO;
	_passedPlayerCharacter = NO;
}

- (void)move {
	if (UI_USER_INTERFACE_IDIOM() == UIUserInterfaceIdiomPad) {
		self.physicsBody.velocity = CGVectorMake(-abRand(450, 600), 0);
	}
	else {
		self.physicsBody.velocity = CGVectorMake(-abRand(200, 300), 0);
	}
}

- (void)addToScene:(ABGameScene *)scene atPosition:(CGPoint)position {
	[super addToScene:scene atPosition:position];
	[self move];
}

- (void)update:(NSTimeInterval)delta {
	if (self.position.x + self.size.width / 2 < 0 || self.position.y + self.size.height / 2 < 0) {
		[self removeFromParent];
	}
}

static BOOL collidedWithCloud = NO;
- (void)collideWith:(SKPhysicsBody *)other {
	if (other.categoryBitMask == ABColliderTypeBird) {
		ABNormalBird *bird = (ABNormalBird *)other.node;
		[bird applyDamage:1];
	}
	else if (other.categoryBitMask == ABColliderTypeCloud) {
		if (!collidedWithCloud) {
			[self runAction:[SKAction fadeAlphaTo:abRand(.3, .9) duration:1]];
			collidedWithCloud = YES;
		}
	}
}

- (void)performDeath {
	self.status = ABGameCharacterStatusDying;
	[self runAction:[SKAction fadeAlphaTo:0 duration:.2] completion: ^{
	    [self removeFromParent];
	    self.status = ABGameCharacterStatusDead;
	}];
}

+ (void)loadSharedFrames {
	static dispatch_once_t onceToken;
	dispatch_once(&onceToken, ^{
	    sharedCloudTexture = [[SKTextureAtlas atlasNamed:@"sprites"] textureNamed:@"Cloud"];
	});
}

static SKTexture *sharedCloudTexture = nil;
- (SKTexture *)cloudTexture {
	return sharedCloudTexture;
}

@end

File: Amazing Bird/Classes/Sprites/ABCloud.swift
import SpriteKit

class ABCloud: ABGameCharacter {
    private static let isPad = UIDevice.current.userInterfaceIdiom == .pad

    private var collidedWithCloud = false
    var hasPassedPlayerCharacter = false

    // MARK: - Init

    init(position: CGPoint) {
        let texture = SKTextureAtlas(named: "sprites").textureNamed("Cloud")
        super.init(texture: texture, position: position)
        name = "cloud"
    }

    required init?(coder: NSCoder) { fatalError("init(coder:) not supported") }

    // MARK: - Physics

    override func configurePhysicsBody() {
        let offsetX = size.width  * anchorPoint.x
        let offsetY = size.height * anchorPoint.y

        let path = CGMutablePath()
        if ABCloud.isPad {
            path.move(to:    CGPoint(x:  17 - offsetX, y: 85 - offsetY))
            path.addLine(to: CGPoint(x: 128 - offsetX, y: 91 - offsetY))
            path.addLine(to: CGPoint(x: 147 - offsetX, y: 29 - offsetY))
            path.addLine(to: CGPoint(x: 101 - offsetX, y:  2 - offsetY))
            path.addLine(to: CGPoint(x:   6 - offsetX, y: 21 - offsetY))
        } else {
            path.move(to:    CGPoint(x:  4 - offsetX, y: 45 - offsetY))
            path.addLine(to: CGPoint(x: 70 - offsetX, y: 53 - offsetY))
            path.addLine(to: CGPoint(x: 83 - offsetX, y: 14 - offsetY))
            path.addLine(to: CGPoint(x: 46 - offsetX, y:  1 - offsetY))
            path.addLine(to: CGPoint(x:  2 - offsetX, y: 14 - offsetY))
        }
        path.closeSubpath()

        physicsBody = SKPhysicsBody(polygonFrom: path)
        physicsBody?.affectedByGravity  = false
        physicsBody?.linearDamping      = 0
        physicsBody?.allowsRotation     = false
        physicsBody?.mass               = 1e1
        physicsBody?.categoryBitMask    = ABColliderType.cloud
        physicsBody?.collisionBitMask   = ABColliderType.bird
        physicsBody?.contactTestBitMask = ABColliderType.bird | ABColliderType.cloud
    }

    // MARK: - Pool reset

    override func reset() {
        super.reset()
        collidedWithCloud = false
        hasPassedPlayerCharacter = false
    }

    // MARK: - Movement

    private func startMoving() {
        let speed = ABCloud.isPad
            ? CGFloat.random(in: 520...650)
            : CGFloat.random(in: 260...350)
        physicsBody?.velocity = CGVector(dx: -speed, dy: 0)
    }

    override func addToScene(_ scene: ABGameScene, atPosition position: CGPoint) {
        super.addToScene(scene, atPosition: position)
        startMoving()
    }

    // MARK: - Update / collision

    override func update(_ delta: TimeInterval) {
        if position.x + size.width / 2 < 0 || position.y + size.height / 2 < 0 {
            removeFromParent()
        }
    }

    override func collideWith(_ body: SKPhysicsBody) {
        if body.categoryBitMask == ABColliderType.bird {
            (body.node as? ABNormalBird)?.applyDamage(1)
        } else if body.categoryBitMask == ABColliderType.cloud, !collidedWithCloud {
            run(SKAction.fadeAlpha(to: CGFloat.random(in: 0.3...0.9), duration: 1))
            collidedWithCloud = true
        }
    }

    // MARK: - Death

    override func performDeath() {
        status = .dying
        run(SKAction.fadeAlpha(to: 0, duration: 0.2)) { [weak self] in
            self?.removeFromParent()
            self?.status = .dead
        }
    }
}
